internal/sftp/goclient_test.go

```diff
@@ -0,0 +1,288 @@
+package sftp_test
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/gliderlabs/ssh"
+	gosftp "github.com/pkg/sftp"
+	gossh "golang.org/x/crypto/ssh"
+	"gotest.tools/v3/assert"
+	tfs "gotest.tools/v3/fs"
+
+	"github.com/artefactual-sdps/enduro/internal/sftp"
+)
+
+// ServerAddress is the test SFTP server address.
+const serverAddress = "127.0.0.1:2222"
+
+// PubkeyHandler handles checking the client's public key against the keys in
+// the authorized_keys file.
+func pubkeyHandler(ctx ssh.Context, key ssh.PublicKey) bool {
+	file, err := os.Open("./testdata/authorized_keys")
+	if err != nil {
+		log.Fatalln("SSH: couldn't open authorized_keys file.")
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		allowed, _, _, _, err := ssh.ParseAuthorizedKey([]byte(scanner.Text()))
+		if err != nil {
+			log.Fatalln("SSH: couldn't parse authorized key.")
+		}
+		if ssh.KeysEqual(key, allowed) {
+			return true
+		}
+	}
+
+	log.Println("SSH: unknown key provided.")
+	return false
+}
```

Comment:
This is not testing the code, it's just set up for the tests.  I believe the tests should not run if server setup fails because otherwise you will get false positives from the client tests (e.g. "ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain")

Reply:
To be clear, this is killing the test binary that Go builds for the entire package. It may prevent other tests from running even if they're defined in other files, e.g. tests in another_impl_test.go.

Reply:
Ohhh, yeah that sounds bad. :(  

I don't think returning false is a good alternative though, because the error message will suggest that the private key provided is unknown, instead of reporting that the authorized_keys file is missing or invalid. :(  Maybe the best thing to do is move all of the authorized_keys file processing out of the pubkeyHandler and handle those errors with the `testing.T` exit methods. :shrug:

Reply:
It turns out that Uber's guide also mentions this issue.

https://github.com/uber-go/guide/blob/master/style.md#exit-in-main